idioms/ffi-accepting-strings.md

```diff
@@ -0,0 +1,80 @@
+# Accepting Strings
+
+## Description
+
+When accepting strings via FFI through pointers, there are two principles that should be followed:
+1. Keep foreign strings "borrowed", rather than copying them directly.
+2. Minimize `unsafe` code during the conversion.
+
+## Motivation
+
+Rust has built-in support for C-style strings with its `CString` and `CStr` types. However, there are different approaches one can take with strings that are being accepted from a foreign caller of a Rust function.
+
+The best practice is simple: use `CStr` in such a way as to minimize unsafe code, and created a borrowed slice. If an owned String is needed, call `to_string()` on the string slice.
+
+## Code Example
+
+```rust
+#[no_mangle]
+pub extern "C" fn mylib_log(msg: *const libc::c_char, level: libc::c_int) {
+    let level: crate::LogLevel = match level { /* ... */ };
+    let msg_str: &str = unsafe {
+        // SAFETY: accessing raw pointers expected to live for the call, and creating a shared
+        // reference that does not outlive the current stack frame.
+        match std::ffi::CStr::from_ptr(path).to_str() {
+            Ok(s) => s,
+            Err(e) => {
+                crate::log_error("FFI string conversion failed");
+                return;
+            }
+        }
+    };
+    crate::log(msg_str, level);
+}
+```
+
+## Advantages
+
+The example is is written to ensure that:
+1. The `unsafe` block is as small as possible.
+1. The pointer with an "untracked" lifetime becomes a "tracked" shared reference
```

Comment:
```suggestion
2. The pointer with an "untracked" lifetime becomes a "tracked" shared reference
```

There are a lot of lists that are like this (1. 1. 1.)

Reply:
This was intentional.

Both `mdbook` and markdown renumber lists in the same block regardless of the actual numbers you put in. Given this, I thought it would be easier to write `1. 1. 1.` than `1. 2. 3.` to avoid future changes that rendered okay ending up `1. 2. 2. 3.` in the source, which might cause some scratching of heads.

That said, I will still fix this if you would like.

Reply:
I have never seen this, nice to know.
Maybe there are some people who reads from markdown directly (for example if you want to contribute).
I will propose the changes, so you only have to accept it :)